File: examples/python/testing/examples/parallel-workflow.py
from aws_durable_execution_sdk_python import DurableContext, durable_execution, durable_step
from aws_durable_execution_sdk_python.types import StepContext
from aws_durable_execution_sdk_python.execution import InvocationStatus
from aws_durable_execution_sdk_python_testing.runner import DurableFunctionTestRunner


@durable_step
def fetch_a(ctx: StepContext) -> str:
    return "data-a"


@durable_step
def fetch_b(ctx: StepContext) -> str:
    return "data-b"


@durable_step
def fetch_c(ctx: StepContext) -> str:
    return "data-c"


@durable_execution
def handler(event, context: DurableContext) -> list:
    results = context.parallel(
        [
            lambda ctx: ctx.step(fetch_a()),
            lambda ctx: ctx.step(fetch_b()),
            lambda ctx: ctx.step(fetch_c()),
        ],
        name="fetch-all",
    )
    return results.get_succeeded()


def test_executes_branches_in_parallel():
    runner = DurableFunctionTestRunner(handler=handler)
    with runner:
        result = runner.run(timeout=10)

    assert result.status is InvocationStatus.SUCCEEDED

File: examples/typescript/testing/examples/parallel-workflow.ts
import { withDurableExecution, DurableContext } from "@aws/durable-execution-sdk-js";
import { LocalDurableTestRunner } from "@aws/durable-execution-sdk-js-testing";
import { ExecutionStatus } from "@aws-sdk/client-lambda";

const handler = withDurableExecution(async (event: unknown, context: DurableContext) => {
  const results = await context.parallel("fetch-all", [
    async (ctx) => await ctx.step("fetch-a", () => "data-a"),
    async (ctx) => await ctx.step("fetch-b", () => "data-b"),
    async (ctx) => await ctx.step("fetch-c", () => "data-c"),
  ]);
  return results.getSucceeded();
});

let runner: LocalDurableTestRunner;

beforeAll(async () => {
  await LocalDurableTestRunner.setupTestEnvironment();
});

afterAll(async () => {
  await LocalDurableTestRunner.teardownTestEnvironment();
});

beforeEach(() => {
  runner = new LocalDurableTestRunner({ handlerFunction: handler });
});

it("executes branches in parallel and collects results", async () => {
  const result = await runner.run();

  expect(result.getStatus()).toBe(ExecutionStatus.SUCCEEDED);
  expect(result.getResult()).toEqual(["data-a", "data-b", "data-c"]);
});
